services/ingestion/subscriber.go

```diff
@@ -2,55 +2,232 @@ package ingestion
 	"github.com/onflow/flow-go/fvm/systemcontracts"
 	flowGo "github.com/onflow/flow-go/model/flow"
+	"github.com/rs/zerolog"
 )
 
 type EventSubscriber interface {
-	// Subscribe to relevant events from the provided block height.
-	// Returns a channel with block events and errors,
-	// if subscription fails returns an error as the third value.
-	Subscribe(ctx context.Context, height uint64) (<-chan flow.BlockEvents, <-chan error, error)
+	// Subscribe to EVM events from the provided height, and return a chanel with the events.
+	//
+	// The BlockEvents type will contain an optional error in case
+	// the error happens, the consumer of the chanel should handle it.
+	Subscribe(ctx context.Context, height uint64) <-chan models.BlockEvents
 }
 
+var _ EventSubscriber = &RPCSubscriber{}
+
 type RPCSubscriber struct {
-	client access.Client
+	client *models.CrossSporkClient
 	chain  flowGo.ChainID
+	logger zerolog.Logger
 }
 
-func NewRPCSubscriber(client access.Client, chainID flowGo.ChainID) *RPCSubscriber {
-	return &RPCSubscriber{client: client, chain: chainID}
+func NewRPCSubscriber(client *models.CrossSporkClient, chainID flowGo.ChainID, logger zerolog.Logger) *RPCSubscriber {
+	logger = logger.With().Str("component", "subscriber").Logger()
+	return &RPCSubscriber{client: client, chain: chainID, logger: logger}
 }
 
-func (r *RPCSubscriber) Subscribe(ctx context.Context, height uint64) (<-chan flow.BlockEvents, <-chan error, error) {
-	// define events we subscribe to: A.{evm}.EVM.BlockExecuted and A.{evm}.EVM.TransactionExecuted,
-	// where {evm} is EVM deployed contract address, which depends on the chain ID we configure
-	evmAddress := common.Address(systemcontracts.SystemContractsForChain(r.chain).EVMContract.Address)
-	blockExecutedEvent := common.NewAddressLocation(nil, evmAddress, string(types.EventTypeBlockExecuted)).ID()
-	transactionExecutedEvent := common.NewAddressLocation(nil, evmAddress, string(types.EventTypeTransactionExecuted)).ID()
+// Subscribe will retrieve all the events from the provided height. If the height is from previous
+// sporks, it will first backfill all the events in all the previous sporks, and then continue
+// to listen all new events in the current spork.
+//
+// If error is encountered during backfill the subscription will end and the response chanel will be closed.
+func (r *RPCSubscriber) Subscribe(ctx context.Context, height uint64) <-chan models.BlockEvents {
+	events := make(chan models.BlockEvents)
 
-	filter := flow.EventFilter{
-		EventTypes: []string{
-			blockExecutedEvent,
-			transactionExecutedEvent,
-		},
-	}
+	go func() {
+		defer func() {
+			close(events)
+		}()
+
+		// if the height is from the previous spork, backfill all the events from previous sporks first
+		if r.client.IsPastSpork(height) {
+			r.logger.Info().
+				Uint64("height", height).
+				Msg("height found in previous spork, starting to backfill")
+
+			for ev := range r.backfill(ctx, height) {
+				events <- ev
```

Comment:
The logic looks right in each of these methods, but I think it's worth adding some comments noting that context cancellation is handled in the producer (`subscribe`), and downstream readers are allowed to drain.

Reply:
Good idea